src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,219 @@
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // ensure llvm knows not to put this into immutable memory
+        alloc.mutability = match self.mode {
+            InternMode::StaticMut => Mutability::Mutable,
+            InternMode::Static { frozen: true } => Mutability::Immutable,
+            InternMode::Static { frozen: ref mut frozen @ false } => {
+                // values behind the initial reference are frozen (except if the type undoes this)
+                *frozen = true;
+                Mutability::Mutable
+            },
+            InternMode::Const { frozen: true } => Mutability::Immutable,
+            InternMode::Const { frozen: ref mut frozen @ false } => {
+                // values behind the initial reference are frozen even if the type would undo this.
+                // validation will detect any discrepancies and tell the user that this is not cool.
+                *frozen = true;
+                Mutability::Mutable
+            },
+        };
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Go over all the primitive types
+        let ty = mplace.layout.ty;
+        if let ty::Ref(..) = ty.sty {
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait objects
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            self.intern(vtable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if mplace.ptr.is_ptr() {
+                match self.intern(mplace.ptr.to_ptr()?)? {
```

Comment:
Instead of `if` followed by `match`, a `if let Ok(ptr) = mplace.ptr.to_ptr()` seems nicer.